Improve top-level docs

jyn514 opened the pull request:
See a detailed explanation in the commit messages. This is a companion PR to https://github.com/rust-lang/rustc-dev-guide/pull/1528.

* Link to other resources instead of inlining their information
* Remove ancient and outdated reference to `config.mk`
* Suggest `profile = "user"` in the README
* Add detail about dependencies from the dev-guide
* Link to CONTRIBUTING.md instead of rustc-dev-guide in the main readme
* Link to `std-dev-guide` in CONTRIBUTING.md

jyn514 commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
+If building LLVM from source (the default), you'll need additional tools:
+
+* `g++`, `clang++`, or MSVC with versions listed on
+  [LLVM's documentation](https://llvm.org/docs/GettingStarted.html#host-c-toolchain-both-compiler-and-standard-library)
```
This changed from the dev-guide; that used a pinned release (LLVM 13.0) but this uses the latest docs. I changed it because I think this README is more likely to get out of date, it doesn't have `<!-- date-check -->` like the dev-guide.

jyn514 commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
    * `git`
-   * `ssl` which comes in `libssl-dev` or `openssl-devel`
+   * OpenSSL (`libssl-dev` or `openssl-devel` on most Unix distros)
+   * XZ (`liblzma-dev` on most distros)
+   * `libffi-dev`
+   * `curl` (not needed on Windows)
    * `pkg-config` if you are compiling on Linux and targeting Linux
+   * `libiconv` (already included with glibc on Debian-based distros)
```
XZ, libffi, and libiconv are new in this list (not from the dev-guide). I found them by looking at the nix build recipe: https://github.com/NixOS/nixpkgs/blob/5528350186a9e826588cee1329640899ca44a0cf/pkgs/development/compilers/rust/rustc.nix#L1-L16
`iconv` is bundled with glibc on most distros, but miri is definitely using libffi and llvm and `rust-installer` are using XZ.

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
    * `git`
-   * `ssl` which comes in `libssl-dev` or `openssl-devel`
+   * OpenSSL (`libssl-dev` or `openssl-devel` on most Unix distros)
+   * XZ (`liblzma-dev` on most distros)
```
Do we really require libssl-dev and liblzma-dev?

For liblzma-dev, I would expect that the build script would just build it if it's not installed (https://docs.rs/crate/lzma-sys/0.1.20/source/build.rs). So it shouldn't be necessary.

For libssl-dev, I guess it's required for Cargo + bump-stage0, but most people won't be building Cargo, and it can be a bit of a painful dependency. So I think my preference for this list is to drop it or at least make clear that it's not strictly required.

```
$ cargo tree -i openssl-sys
openssl-sys v0.9.72
├── curl v0.4.44
│   ├── bump-stage0 v0.1.0 (/home/mark/Edit/rust/src/tools/bump-stage0)
│   ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   ├── crates-io v0.35.0 (/home/mark/Edit/rust/src/tools/cargo/crates/crates-io)
│   │   ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   │   └── cargo-test-support v0.1.0 (/home/mark/Edit/rust/src/tools/cargo/crates/cargo-test-support)
│   │       [dev-dependencies]
│   │       └── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   └── git2-curl v0.16.0
│       └── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
├── curl-sys v0.4.59+curl-7.86.0
│   ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   └── curl v0.4.44 (*)
├── git2 v0.15.0
│   ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   ├── cargo-test-support v0.1.0 (/home/mark/Edit/rust/src/tools/cargo/crates/cargo-test-support) (*)
│   └── git2-curl v0.16.0 (*)
├── libgit2-sys v0.14.0+1.5.0
│   ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
│   └── git2 v0.15.0 (*)
├── libssh2-sys v0.2.23
│   └── libgit2-sys v0.14.0+1.5.0 (*)
└── openssl v0.10.38
    └── crypto-hash v0.3.4
        └── cargo-util v0.2.3 (/home/mark/Edit/rust/src/tools/cargo/crates/cargo-util)
            ├── cargo v0.68.0 (/home/mark/Edit/rust/src/tools/cargo)
            └── cargo-test-support v0.1.0 (/home/mark/Edit/rust/src/tools/cargo/crates/cargo-test-support) (*)
```

jyn514 replied:
Oh huh, I guess we're only using it for rust-installer. I thought we linked it in rustc but it turns out I was thinking of zlib, not xz. https://github.com/rust-lang/rust/issues/74420

I'll remove liblzma and move libssl to its own section.

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
+
+Make sure you have installed the dependencies:
+
+   * `python` 3 or 2.7 (under the name python; python2 or python3 will not work)
```
Is it really true that python2 or python3 won't work? What creates that requirement?

jyn514 replied:
`./x.py` requires exactly `python`. Running `python3 x.py` still works as usual. How about suggesting `python x.py` in the readme and removing the parentheses about python2/python3?

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
-   * `ssl` which comes in `libssl-dev` or `openssl-devel`
+   * OpenSSL (`libssl-dev` or `openssl-devel` on most Unix distros)
+   * XZ (`liblzma-dev` on most distros)
+   * `libffi-dev`
```
This should only be needed on Linux (miri depends on it in target_os = linux), and even there I'm not sure it's required:

> This crate depends on [the libffi-sys crate](https://crates.io/crates/libffi-sys/), which by default attempts to build its own version of the C libffi library. In order to use your system’s C libffi instead, enable this crate’s system feature in your Cargo.toml:

Miri doesn't enable the system feature.

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
    * `pkg-config` if you are compiling on Linux and targeting Linux
+   * `libiconv` (already included with glibc on Debian-based distros)
+
+If building LLVM from source (the default), you'll need additional tools:
```
I thought the default is fetching from CI? At minimum, it seems worth calling that out as a possibility...

jyn514 replied:
Good catch, thanks. I changed it to remove any mention of a default, since we've been changing it a lot recently and I doubt people will remember to update the readme on changes.

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
+
+Make sure you have installed the dependencies:
 
-   * `g++` 5.1 or later or `clang++` 3.5 or later
```
I think we probably have C/C++ dependencies somewhere in the compiler tree? At minimum, these are needed for any Rust compilation as we typically invoke the linker through cc, right? (Not sure about Windows, but I believe this is the case on Ubuntu at least).

jyn514 replied:
We need at least cc, but not a full C++ compiler. I can say that explicitly if you like, but I think most people assume they need at least a C compiler for most things.

Mark-Simulacrum replied:
Hm. I guess I would probably not call that an implicit assumption we can make; I think we should say it. Folks coming from languages like Node or Python might not realize that they need extra compilers to work on std, for example.

(And long-term I think we should try to remove this system dependency, e.g., by including clang in our artifacts. But that's a very separate conversation).

Mark-Simulacrum commented on README.md:
```diff
@@ -48,20 +44,36 @@ by running it with the `--help` flag or reading the [rustc dev guide][rustcguide
+* `libstdc++-static` may be required on some Linux distributions such as Fedora and Ubuntu
 
-2. Clone the [source] with `git`:
+On tier 1 platforms, you can also choose to download LLVM by setting `llvm.download-ci-llvm = true`.
```
```suggestion
On tier 1 or tier 2 with host tools platforms, you can also choose to download LLVM by setting `llvm.download-ci-llvm = true`.
```

Right?

Mark-Simulacrum commented:
r=me with last comment resolved and commits squashed (at least a bit)

jyn514 committed:
Suggest `profile = "user"` in the README

This sets several useful defaults, like `extended = true`, and gives us a path forward for changing
global defaults without breaking distros.

jyn514 committed:
Add detail about dependencies from the dev-guide

The goal is to remove this altogether from the dev-guide once this PR is merged.

jyn514 committed:
Link to other resources instead of inlining their information

The other places are more accurate and up-to-date.

- Link to `std-dev-guide` in CONTRIBUTING.md

  Thom and Mara said the guide is in reasonably good shape, and it's tailored more closely to people working on the standard library.

- Link to CONTRIBUTING.md instead of rustc-dev-guide in the main readme

  CONTRIBUTING.md has more information and also links the std-dev-guide.

- Link to forge for the list of tested platforms; the one in the readme
  was hopelessly out of date.

jyn514 committed:
Remove ancient and outdated references to `config.mk`

jyn514 commented:
@bors r=Mark-Simulacrum rollup

bors commented:
:pushpin: Commit 70a0e0a2779f04999d0f0e40f93e1697c37ec042 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 70a0e0a2779f04999d0f0e40f93e1697c37ec042 -->
<!-- homu: {"type":"Approved","sha":"70a0e0a2779f04999d0f0e40f93e1697c37ec042","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
